osu.Server.Spectator/Hubs/MultiplayerHub.cs

```diff
@@ -0,0 +1,388 @@
+            }
+        }
+
+        public async Task StartMatch()
+        {
+            var room = await getLocalUserRoom();
+
+            ensureIsHost(room);
+
+            using (room.LockForUpdate())
+            {
+                if (room.State != MultiplayerRoomState.Open)
+                    throw new InvalidStateException("Can't start match when already in a running state.");
+
+                var readyUsers = room.Users.Where(u => u.State == MultiplayerUserState.Ready).ToArray();
+
+                if (readyUsers.Length == 0)
+                    throw new InvalidStateException("Can't start match when no users are ready.");
+
+                if (room.Host != null && room.Host.State != MultiplayerUserState.Ready)
+                    throw new InvalidStateException("Can't start match when the host is not ready.");
+
+                await changeRoomState(room, MultiplayerRoomState.WaitingForLoad);
+
+                foreach (var u in readyUsers)
+                    await changeAndBroadcastUserState(room, u, MultiplayerUserState.WaitingForLoad);
+
+                await Clients.Group(GetGroupId(room.RoomID, true)).LoadRequested();
+            }
+        }
+
+        public async Task ChangeSettings(MultiplayerRoomSettings settings)
+        {
+            var room = await getLocalUserRoom();
+
+            using (room.LockForUpdate())
+            {
+                if (room.State != MultiplayerRoomState.Open)
+                    throw new InvalidStateException("Attempted to change settings while game is active");
+
+                ensureIsHost(room);
+
+                room.Settings = settings;
+                await Clients.Group(GetGroupId(room.RoomID)).SettingsChanged(settings);
+            }
+        }
+
+        protected override Task OnDisconnectedAsync(Exception exception, MultiplayerClientState? state)
+        {
+            if (state != null)
+                return LeaveRoom();
+
+            return base.OnDisconnectedAsync(exception, state);
+        }
+
+        /// <summary>
+        /// Get the group ID to be used for multiplayer messaging.
+        /// </summary>
+        /// <param name="roomId">The databased room ID.</param>
+        /// <param name="gameplay">Whether the group ID should be for active gameplay, or room control messages.</param>
+        public static string GetGroupId(long roomId, bool gameplay = false) => $"room:{roomId}:{gameplay}";
+
+        /// <summary>
+        /// Should be called when user states change, to check whether the new overall room state can trigger a room-level state change.
+        /// </summary>
+        private async Task updateRoomStateIfRequired(MultiplayerRoom room)
+        {
+            //check whether a room state change is required.
+            switch (room.State)
+            {
+                case MultiplayerRoomState.WaitingForLoad:
+                    if (room.Users.All(u => u.State != MultiplayerUserState.WaitingForLoad))
```

Comment:
From what I gathered, the user can always back out to `Idle`, in light of which this should probably be checking

```suggestion
                    if (room.Users.All(u => u.State == MultiplayerUserState.Loaded))
```

and have a special case for all users re-entering `Idle`. Here's a failing test case:

```csharp
        [Fact]
        public async Task AllUsersBackingOutFromLoadCancelsTransitionToPlay()
        {
            var room = await hub.JoinRoom(room_id);

            setUserContext(mockContextUser2);
            await hub.JoinRoom(room_id);

            await hub.ChangeState(MultiplayerUserState.Ready);

            setUserContext(mockContextUser1);
            await hub.ChangeState(MultiplayerUserState.Ready);
            await hub.StartMatch();

            Assert.Equal(MultiplayerRoomState.WaitingForLoad, room.State);

            await hub.ChangeState(MultiplayerUserState.Idle);
            setUserContext(mockContextUser2);
            await hub.ChangeState(MultiplayerUserState.Idle);

            Assert.Equal(MultiplayerRoomState.Open, room.State);
        }
```